src/libs/FileSorter.cpp

```diff
@@ -0,0 +1,204 @@
+/*
+ * FileSorter.cpp
+ *
+ *  Created on: Sep 29, 2016
+ *      Author: mmoore
+ */
+
+#include "FileSorter.h"
+#include "DirHandle.h"
+#include "FileStream.h"
+
+#include <new>
+#include <malloc.h>
+#include <string.h>
+
+
+FileSorter::FileSorter(string dir_path, __file_sort_filter_fn_t callback)
+{
+    this->directory_path = dir_path;
+    this->error = false;
+    this->file_info_array = NULL;
+    this->file_count = 0;
+    this->total_file_count = 0;
+    this->file_index = -1;
+    this->filter_callback = callback;
+
+    open_directory(dir_path);
+}
+
+FileSorter::~FileSorter()
+{
+    delete_file_info_array();
+}
+
+int FileSorter::compare_dir(file_info_t* file_a, file_info_t* file_b)
+{
+    return strcmp(file_a->file_name, file_b->file_name);
+}
+
+size_t FileSorter::count_files_in_folder(string dir_path)
+{
+    struct dirent *p;
+    uint16_t count = 0;
+    DIR* folder = opendir(dir_path.c_str());
+    if ( folder != NULL ) {
+        while ( (p = readdir(folder)) != NULL ) {
+            if ( (this->filter_callback == NULL) || this->filter_callback(p) ) {
+                count++;
+            }
+        }
+        closedir(folder);
+        return count;
+    }
+    return 0;
+}
+
+void FileSorter::sort_directory(void)
+{
+    // sort the array.
+    qsort(this->file_info_array, this->file_count, sizeof(file_info_t), (__compar_fn_t)compare_dir);
+    this->file_index = -1;
+}
+
+void FileSorter::delete_file_info_array(void)
+{
+    if ( this->file_info_array != NULL) {
+        // file names were allocated previously with strdup - free them now.
+        for ( size_t i=0; i<this->file_count; i++ ) {
+            free(this->file_info_array[i].file_name);
+        }
+
+        // free the array.
+        delete[] this->file_info_array;
+    }
+}
+
+void FileSorter::open_directory(string dir_path)
+{
+    // reset all member state variables.
+    this->directory_path = dir_path;
+    this->error = false;
+    this->file_count = 0;
+    this->total_file_count = 0;
+    this->file_index = -1;
+    delete_file_info_array();
+
```

Comment:
Hey John,

I pushed changes to fix the memory leak and address your review notes below.  I've not yet addressed items 1-3 in this comment, but I'm happy to do it when I have opportunity.  I didn't originally consider items 2 and 3 because I really had not considered submitting these changes at all.  I made the modification for my own use, and later thought others might find it helpful as well.  Items 2 and 3 definitely make sense in the context of a wider audience.  I'll address them when I have some time and push them up.  As for item 1, my reasons for putting it on the heap was that it is an 'if-available' feature.  If there's enough memory to do the sort, then do it.  Otherwise, no sorting, so I wanted to completely clean up resources when done.  I suppose I can create it as a member variable and add a method to the `FileSorter` to clean up resources on exit.  I'm not certain that would get it off the heap though.  Wouldn't the fact that the fact that the `FileScreen` itself is created on the heap cause all its members to be on the heap as well:
```
this->file_screen    = (new FileScreen()            )->set_parent(this);
```